src/libfuncs/snapshot_take.rs

```diff
@@ -1,3 +1,45 @@
 //! # Snapshot taking libfuncs
 //!
 //! TODO
+
+use super::{LibfuncBuilder, LibfuncHelper};
+use crate::{metadata::MetadataStorage, types::TypeBuilder};
+use cairo_lang_sierra::{
+    extensions::{lib_func::SignatureOnlyConcreteLibfunc, GenericLibfunc, GenericType},
+    program_registry::ProgramRegistry,
+};
+use melior::{
+    ir::{Block, Location},
+    Context,
+};
+
+/// Generate MLIR operations for the `snapshot_take` libfunc.
+pub fn build<'ctx, 'this, TType, TLibfunc>(
+    _context: &'ctx Context,
+    _registry: &ProgramRegistry<TType, TLibfunc>,
+    entry: &'this Block<'ctx>,
+    location: Location<'ctx>,
+    helper: &LibfuncHelper<'ctx, 'this>,
+    _metadata: &mut MetadataStorage,
+    _info: &SignatureOnlyConcreteLibfunc,
+) -> Result<(), std::convert::Infallible>
+where
+    TType: GenericType,
+    TLibfunc: GenericLibfunc,
+    <TType as GenericType>::Concrete: TypeBuilder,
+    <TLibfunc as GenericLibfunc>::Concrete: LibfuncBuilder,
+{
+    // TODO: Should this act like dup like it does now? or are there special requirements. So far it seems to work.
+    // TODO: Handle non-trivially-copyable types (ex. arrays) and maybe update docs.
```

Comment:
Judging by the name and the fact that there are references (and are different from snapshots) maybe a `Snapshot<T>` is roughly equivalent to a `Cow<T>` where it provides a "copy" of a value that is not really copied until used/taken, but doesn't get updated if the original changes?